x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/info/FieldSelection.java

```diff
@@ -0,0 +1,184 @@
+
+    private static final ParseField NAME = new ParseField("name");
+    private static final ParseField MAPPING_TYPES = new ParseField("mapping_types");
+    private static final ParseField IS_INCLUDED = new ParseField("is_included");
+    private static final ParseField IS_REQUIRED = new ParseField("is_required");
+    private static final ParseField FEATURE_TYPE = new ParseField("feature_type");
+    private static final ParseField REASON = new ParseField("reason");
+
+    public enum FeatureType {
+        CATEGORICAL, NUMERICAL;
+
+        public static FeatureType fromString(String value) {
+            return FeatureType.valueOf(value.toUpperCase(Locale.ROOT));
+        }
+
+        @Override
+        public String toString() {
+            return name().toLowerCase(Locale.ROOT);
+        }
+    }
+
+    public static ConstructingObjectParser<FieldSelection, Void> PARSER = new ConstructingObjectParser<>("field_selection",
+        a -> new FieldSelection((String) a[0], new HashSet<>((List<String>) a[1]), (boolean) a[2], (boolean) a[3], (FeatureType) a[4],
+            (String) a[5]));
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), NAME);
+        PARSER.declareStringArray(ConstructingObjectParser.constructorArg(), MAPPING_TYPES);
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), IS_INCLUDED);
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), IS_REQUIRED);
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(), p -> {
+            if (p.currentToken() == XContentParser.Token.VALUE_STRING) {
+                return FeatureType.fromString(p.text());
+            }
+            throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
+        }, FEATURE_TYPE, ObjectParser.ValueType.STRING);
+        PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), REASON);
+    }
+
+    private final String name;
+    private final Set<String> mappingTypes;
+    private final boolean isIncluded;
+    private final boolean isRequired;
+    private final FeatureType featureType;
+    private final String reason;
+
+    public static FieldSelection included(String name, Set<String> mappingTypes, boolean isRequired, FeatureType featureType) {
+        return new FieldSelection(name, mappingTypes, true, isRequired, featureType, null);
+    }
+
+    public static FieldSelection excluded(String name, Set<String> mappingTypes, String reason) {
+        return new FieldSelection(name, mappingTypes, false, false, null, reason);
+    }
+
+    FieldSelection(String name, Set<String> mappingTypes, boolean isIncluded, boolean isRequired, @Nullable FeatureType featureType,
+                           @Nullable String reason) {
+        this.name = Objects.requireNonNull(name);
+        this.mappingTypes = Collections.unmodifiableSet(mappingTypes);
```

Comment:
`Objects.requireNonNull`?
Also, should we use `ExceptionHelper.requireNonNull` instead?

Reply:
`ExceptionsHelper.requireNonNull` makes sense for user facing code. We thus use it when we parse request objects. In this case we create this ourselves `Objects.requireNonNull` should be enough.